libwnbd/wnbd_ioctl.cpp

```diff
@@ -139,6 +144,358 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
     return Status;
 }
 
+DWORD RemoveWnbdInf(_In_ LPCTSTR InfName)
+{
+    DWORD Status = 0;
+    UINT ErrorLine;
+
+    HINF HandleInf = SetupOpenInfFile(InfName, NULL, INF_STYLE_WIN4, &ErrorLine);
+    if (HandleInf == INVALID_HANDLE_VALUE) {
+        Status = GetLastError();
+        LogError("SetupOpenInfFile failed with "
+            "error: %d, at line. Error message: %s",
+            ErrorLine, Status, win32_strerror(Status).c_str());
+        goto failed;
+    }
+
+    INFCONTEXT Context;
+    if (!SetupFindFirstLine(HandleInf, INFSTR_SECT_VERSION, INFSTR_KEY_CATALOGFILE, &Context)) {
+        Status = GetLastError();
+        LogError("SetupFindFirstLine failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto failed;
+    }
+
+    TCHAR InfData[MAX_INF_STRING_LENGTH];
+    if (!SetupGetStringField(&Context, 1, InfData, ARRAYSIZE(InfData), NULL)) {
+        Status = GetLastError();
+        LogError("SetupGetStringField failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto failed;
+    }
+
+    /* Match the OEM inf file based on the catalog string wnbd.cat */
+    if (!wcscmp(InfData, L"wnbd.cat")) {
+        std::wstring SearchString(InfName);
+        if (!SetupUninstallOEMInf(
+            SearchString.substr(SearchString.find_last_of(L"\\") + 1).c_str(), SUOI_FORCEDELETE, 0)) {
+            Status = GetLastError();
+            LogError("SetupUninstallOEMInfA failed with "
+                "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        }
+    }
+
+failed:
+    if (HandleInf != INVALID_HANDLE_VALUE) {
+        SetupCloseInfFile(HandleInf);
+    }
+
+    return Status;
+}
+
+/* Cycle through all OEM inf files and try a best effort mode to remove all
+ * files which include the string wnbd.cat */
+DWORD CleanDrivers()
+{
+    DWORD Status = 0;
+    TCHAR OemName[MAX_PATH];
+    HANDLE DirHandle = INVALID_HANDLE_VALUE;
+    WIN32_FIND_DATA OemFindData;
+
+    if (!GetWindowsDirectory(OemName, ARRAYSIZE(OemName))) {
+        Status = GetLastError();
+        LogError("GetWindowsDirectory failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        return Status;
+    }
+    if (wcscat_s(OemName, ARRAYSIZE(OemName), L"\\INF\\OEM*.INF")) {
+        LogError("Couldn't process path: %ls.", OemName);
+        return ERROR_BAD_PATHNAME;
+    }
+    DirHandle = FindFirstFile(OemName, &OemFindData);
+    if (DirHandle == INVALID_HANDLE_VALUE) {
+        /* No OEM files */
+        return Status;
+    }
+
+    do {
+        Status = RemoveWnbdInf(OemFindData.cFileName);
```

Comment:
We should preserve the first or the last non-zero status. If out of 5 files we fail to remove one, the status should be non-zero.

Reply:
My bad, I was expecting it to loop over all drivers and return an error afterwards. Now that I see that we're breaking the loop when hitting an error, I guess it's fine.